docs/src/constraints.md

```diff
@@ -1,5 +1,514 @@
+is similar to the [`@variable`](@ref) macro. Here is an example of how to add
+the constraint ``2x \le 1`` to a JuMP model:
+```jldoctest con1; setup = :(model = Model(); @variable(model, x))
+julia> @constraint(model, con, 2x <= 1)
+con : 2 x <= 1.0
+```
+Wasn't that easy! Let's unpack what happened, because just like
+[`@variable`](@ref) there are a few subtle things going on.
+ 1. The mathematical constraint ``2x \le 1`` was added to the model.
+ 2. A JuMP called `con` was created that acts as a reference to the
+    constraint.
+ 3. JuMP set the *name* of the constraint to `"con"`.
+
+Just like the Julia variables created in [`@variable`](@ref), `con` can be bound
+to a different value. For example:
+```jldoctest con1
+julia> con
+con : 2 x <= 1.0
+
+julia> con = 1
+1
+
+julia> con
+1
+```
+However, the reference can be retrieved by querying the model using the symbolic
+name:
+```jldoctest con1
+julia> con = model[:con]
+con : 2 x <= 1.0
+
+julia> con
+con : 2 x <= 1.0
+```
+Because the named variables and constraints are stored in the same namespace,
+creating a constraint with the same name as a variable or an existing constraint
+will result in an error. To overcome this limitation, it is possible to create
+anonymous constraints, just like it is possible to create
+[Anonymous JuMP variables](@ref). This is done by dropping the second argument
+to [`@constraint`](@ref):
+```jldoctest con1
+julia> con = @constraint(model, 2x <= 1)
+2 x <= 1.0
+```
+
+It is also possible use different comparison operators (e.g., `>=` and `==`) to
+create the following types of constraints:
+```jldoctest con1
+julia> @constraint(model, 2x >= 1)
+2 x >= 1.0
+
+julia> @constraint(model, 2x == 1)
+2 x = 1.0
+
+julia> @constraint(model, 1 <= 2x <= 3)
+2 x ∈ [1.0, 3.0]
+```
+
+Note that JuMP normalizes the constraints given by the user by moving all of the
+terms containing variables to the left-hand side, and all of the constant terms
+to the right-hand side. Thus, we get:
+```jldoctest; setup=:(model=Model(); @variable(model, x))
+julia> @constraint(model, 2x + 1 <= 4x + 4)
+-2 x <= 3.0
+```
+
+## [Duality](@id constraint_duality)
+
+JuMP adopts the notion of duality from MathOptInterface. Roughly speaking, the
+dual variable associated with a constraint can be viewed as the decrease in the
+value of the objective given an infinitesimal relaxation of the constraint. For
+example, in a linear constraint, this relaxation is equivalent to `2x <= 1 + δ`
+and `2x >= 1 - δ`. If the constraint is an equality constraint, it depends on
+which direction is binding.
+
+Notably, this definition is independent of the objective sense. That is, the
+dual associated with a constraint is the same when the objective is a
+maximization and when it is a minimization. **This is different to traditional
+duality in linear programming.**
+
+The dual value associated with a constraint can be accessed using the
```

Comment:
... associated with a constraint in the most recent solution. Should mention `has_dual` to see if you can query the dual.